File: frontend/src/api/invest.ts
import { apiGet } from "./client";

export interface DCAProjectionPoint {
  month: string;
  projected_krw: number;
  actual_krw: number | null;
  achievement_pct: number | null;
  has_data: boolean;
}

export interface YearlyAchievement {
  year: number;
  projected_year_end_krw: number;
  actual_year_end_krw: number | null;
  achievement_pct: number | null;
  has_data: boolean;
}

export interface GoalTimeline {
  months_to_goal: number | null;
  expected_goal_date: string | null;
  actual_expected_goal_date: string | null;
  current_progress_pct: number | null;
  on_track: boolean | null;
  lead_lag_months: number | null;
}

export interface DCASettings {
  monthly_deposit_amount: number | null;
  goal_annual_return_pct: number | null;
  goal_amount: number | null;
  goal_start_date: string | null;
  goal_initial_amount: number | null;
}

export interface DCAAnalysisData {
  settings: DCASettings;
  projection_months: DCAProjectionPoint[];
  yearly_achievements: YearlyAchievement[];
  goal_timeline: GoalTimeline;
  is_configured: boolean;
}

export const fetchDCAAnalysis = () => apiGet<DCAAnalysisData>("/invest/dca-analysis");

export interface MonthlyProjected {
  month: number;
  amount_krw: number;
}

export interface YearlyReceived {
  year: number;
  amount_krw: number;
}

export interface DividendPlanData {
  annual_dividend_goal: number | null;
  estimated_annual_krw: number;
  estimated_monthly_krw: number;
  actual_annual_received_krw: number;
  goal_achievement_pct: number | null;
  monthly_projected: MonthlyProjected[];
  monthly_received: { month: string; amount: number }[];
  yearly_received: YearlyReceived[];
}

export const fetchDividendPlan = () => apiGet<DividendPlanData>("/invest/dividend-plan");

File: backend/app/schemas/invest.py
from pydantic import BaseModel


class DCAProjectionPoint(BaseModel):
    month: str  # "2024-01"
    projected_krw: float
    actual_krw: float | None
    achievement_pct: float | None
    has_data: bool


class YearlyAchievement(BaseModel):
    year: int
    projected_year_end_krw: float
    actual_year_end_krw: float | None
    achievement_pct: float | None
    has_data: bool


class GoalTimeline(BaseModel):
    months_to_goal: int | None
    expected_goal_date: str | None  # "2039-01" — 최초 계획 기준 달성 예정일
    actual_expected_goal_date: str | None = None  # "2037-04" — 현재 실적 기준 달성 예상일
    current_progress_pct: float | None  # 대시보드 goal_achievement_pct와 동일 소스(총자산÷목표금액)
    on_track: bool | None
    lead_lag_months: int | None  # 양수=앞서는 개월, 음수=뒤처지는 개월


class DCASettings(BaseModel):
    monthly_deposit_amount: float | None
    goal_annual_return_pct: float | None
    goal_amount: float | None
    goal_start_date: str | None  # "2024-01-01"
    goal_initial_amount: float | None


class DCAAnalysisResponse(BaseModel):
    settings: DCASettings
    projection_months: list[DCAProjectionPoint]
    yearly_achievements: list[YearlyAchievement]
    goal_timeline: GoalTimeline
    is_configured: bool


class DepositGuideItem(BaseModel):
    annual_return_pct: float
    required_monthly_deposit: float
    required_annual_deposit: float


class GoalFeasibilityPreview(BaseModel):
    required_return_pct: float | None  # None이면 탐색 범위(-90%~500%) 내에서 달성 불가능
    pv: float  # 계산에 사용된 현재 자산(직접 입력값 또는 총자산 스냅샷)
    n_months: int
    note: str | None = None
    # 가정 수익률 프리셋(보수적/중립/공격적)별 필요 월/연 적립액 — monthly_deposit_amount와 무관하게
    # pv·목표금액·기간만으로 계산되는 별도 가이드(목표 달성에 필요한 월 적립액을 스스로 지어내지 않도록)
    deposit_guide: list[DepositGuideItem] = []
